gazelle_haskell_modules/lang.go

```diff
@@ -171,10 +173,57 @@ func (*gazelleHaskellModulesLang) Embeds(r *rule.Rule, from label.Label) []label
 
 func (*gazelleHaskellModulesLang) Resolve(c *config.Config, ix *resolve.RuleIndex, rc *repo.RemoteCache, r *rule.Rule, imports interface{}, from label.Label) {
 	hmc := c.Exts[gazelleHaskellModulesName].(Config)
-	if isNonHaskellModule(r.Kind()) {
-		setNonHaskellModuleDeps(&hmc, c.RepoRoot, ix, r, imports.(*HRuleImportData), from)
-	} else {
-		setHaskellModuleDeps(ix, r, imports.(*HModuleImportData), from)
+	switch expr := imports.(type) {
+	// This happens only when one run gazelle_cabal and gazelle_haskell_modules in one pass as the the gazelle_binary.
+	case gazelle_cabal.ImportData:
+		// We fully generate the rule that gazele_cabal would have produced.
+		gazelle_cabal.RunResolve(c, ix, rc, r, expr, from)
+		// We then create an HRuleImportData from the generated rule.
+		newImports := mkHRuleImportDataFromGazelleCabalResult(c.RepoRoot, r)
+		setNonHaskellModuleDeps(&hmc, c.RepoRoot, ix, r, &newImports, from)
+	default:
+		if isNonHaskellModule(r.Kind()) {
+			setNonHaskellModuleDeps(&hmc, c.RepoRoot, ix, r, imports.(*HRuleImportData), from)
+		} else {
+			setHaskellModuleDeps(ix, r, imports.(*HModuleImportData), from)
+		}
+	}
+}
+
+func mkHRuleImportDataFromGazelleCabalResult(repoRoot string, r *rule.Rule) HRuleImportData {
+	// From the srcs attribute of r,
+	// we generate the Srcs field of the HRuleImportData.
+	var srcs []string
+	oldSrcs := r.Attr("srcs")
+	if oldSrcs != nil {
+		srcsList := gazelle_cabal.ListOfStringExpr(oldSrcs)
+		for _, x := range srcsList {
+			srcs = append(srcs, repoRoot+"/"+x)
+		}
+		// Now that they are stored in the HRuleImportData,
+		// gazelle_haskell_modules will take care of them,
+		// so we can suppress the attribute srcs from r.
+		r.DelAttr("srcs")
+	}
+	// Same thing for the Deps field.
+	deps := map[label.Label]bool{}
+	oldDeps := r.Attr("deps")
+	if oldDeps != nil {
+		depsList := gazelle_cabal.ListOfStringExpr(oldDeps)
+		for _, x := range depsList {
+			depLbl, _ := label.Parse(x)
+			deps[depLbl] = true
+		}
+		r.DelAttr("deps")
+	}
+	return HRuleImportData{
+		Deps: deps,
+		// When a rule comes from a gazelle_cabal run,
+		// it does not contain any relevant information in the field modules
+		// (this field is either non-existent or filled with data from an outdated run),
+		// hence we can safely put an empty map here.
+		Modules: make(map[label.Label]bool),
+		Srcs:    srcs,
```

Comment:
Is this duplicating the code in `addNonHaskellModuleRules` that creates `HRuleImportData` for preexisting non-`haskell_module` rules?

Reply:
Yes, it is. Is there any tips to discover the already implemented functions in our projects? Furthermore, I am not sure I understand clearly what the various arguments of this function represent.

Reply:
> Is there any tips to discover the already implemented functions in our projects?

I don't think so, other than reading the code :)

> Furthermore, I am not sure I understand clearly what the various arguments of this function represent.

Here's an attempt at explaining them:
```
func addNonHaskellModuleRules(
	c *Config,
	pkgRoot string,
	repo string,
	pkg string,
	gen language.GenerateResult,
	rules []*rule.Rule,
) language.GenerateResult
```
* `c` is the same config parameter as given to `Resolve`
* `pkgRoot` root of the project tree
* `repo` is the name of the bazel repo
* `pkg` is the package path inside project tree
* `gen` is the container to which new generated rules are added

There is [this invocation](https://github.com/tweag/gazelle_haskell_modules/blob/60ad6a95e300ab2f9924ef33be6a62908b1592d7/gazelle_haskell_modules/lang.go#L195) that may give further clues on how the arguments are created. Probably we won't need all of these arguments after factoring out the piece that is needed by both `addNonHaskellModuleRules` and `Resolve`.